pkg/execution/state/redis_state/queue_processor.go

```diff
@@ -498,30 +509,124 @@ func (q *queue) worker(ctx context.Context, f osqueue.RunFunc) {
+func (q *queue) shadowWorker(ctx context.Context, qspc chan *QueueShadowPartition) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
 
-		case <-qspc:
-			metrics.ActiveShadowScanerCount(ctx, 1, metrics.CounterOpt{PkgName: pkgName})
+		case shadowPart := <-qspc:
+			err := q.processShadowPartition(ctx, shadowPart, 0)
+			if err != nil {
+				logger.StdlibLogger(ctx).Error("could not scan shadow partition", "err", err, "shadow_part", shadowPart)
+			}
+		}
+	}
+}
+
+func (q *queue) processShadowPartition(ctx context.Context, shadowPart *QueueShadowPartition, continuationCount uint) error {
+	metrics.ActiveShadowScannerCount(ctx, 1, metrics.CounterOpt{PkgName: pkgName})
+	defer metrics.ActiveShadowScannerCount(ctx, -1, metrics.CounterOpt{PkgName: pkgName})
 
-			// TODO:
-			//
-			// - retrieve countinuation counter
-			//
-			// loop:
-			// - extend lease for shadow partition
-			// - iterate through backlogs
-			//   + add item to function partition
-			//
+	// acquire lease for shadow partition
+	leaseID, err := duration(ctx, q.primaryQueueShard.Name, "shadow_partition_lease", q.clock.Now(), func(ctx context.Context) (*ulid.ULID, error) {
+		leaseID, err := q.ShadowPartitionLease(ctx, shadowPart, ShadowPartitionLeaseDuration)
+		return leaseID, err
+	})
+	if err != nil {
+		if errors.Is(err, ErrShadowPartitionAlreadyLeased) {
+			// contention
+			return nil
+		}
+	}
+
+	if leaseID == nil {
+		return fmt.Errorf("missing shadow partition leaseID")
+	}
+
+	extendLeaseCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// keep extending lease while we're processing
+	go func() {
+		for {
+			select {
+			case <-extendLeaseCtx.Done():
+				return
+			case <-time.Tick(ShadowPartitionLeaseDuration / 2):
+				leaseID, err = q.ShadowPartitionExtendLease(ctx, shadowPart, *leaseID, ShadowPartitionLeaseDuration)
+				if err != nil {
+					if errors.Is(err, ErrShadowPartitionAlreadyLeased) || errors.Is(err, ErrShadowPartitionLeaseNotFound) {
+						// contention
+						return
+					}
+					return
+				}
+			}
+		}
+	}()
+
+	// Check if shadow partition cannot be processed (paused/refill disabled, etc.)
+	if shadowPart.PauseRefill {
+		q.removeShadowContinue(ctx, shadowPart, false)
+
+		forceRequeueAt := q.clock.Now().Add(ShadowPartitionRefillPausedRequeueExtension)
+		err = q.ShadowPartitionRequeue(ctx, shadowPart, *leaseID, &forceRequeueAt)
+		if err != nil {
+			return fmt.Errorf("could not requeue shadow partition: %w", err)
+		}
+
+		return nil
+	}
+
+	until := q.clock.Now()
+	limit := ShadowPartitionPeekMaxBacklogs
+	backlogs, err := q.ShadowPartitionPeek(ctx, shadowPart, until, limit)
+	if err != nil {
+		return fmt.Errorf("could not peek backlogs for shadow partition: %w", err)
+	}
+
+	// Sequentially refill backlogs
+	for _, backlog := range backlogs {
+		err := q.BacklogRefill(ctx, backlog, shadowPart)
+		if err != nil {
+			// TODO Handle capacity reached and exit
+			isCapacityReached := false
```

Comment:
I assume there will be a follow up to convert this variable to true?
I think you can return it as part of `BacklogRefill` too so it's not a separate call to the storage system